File: rtls-kmp/src/commonMain/kotlin/com/rtls/kmp/SyncMergeStrategy.kt
package com.rtls.kmp

/**
 * Context passed to merge strategy when resolving a conflict.
 */
data class MergeContext(
    val lastSyncAtMs: Long? = null,
    val serverTimeMs: Long? = null
)

/**
 * Result of resolving a conflict: keep local, keep server, or use a merged item.
 */
sealed class LocationMergeResult {
    data object KeepLocal : LocationMergeResult()
    data object KeepServer : LocationMergeResult()
    data class Use(val point: LocationPoint) : LocationMergeResult()
}

/**
 * Optional. If not provided, default is server-wins for pull (overwrite local with server).
 */
interface LocationMergeStrategy {
    fun resolve(
        local: LocationPoint?,
        server: LocationPoint,
        context: MergeContext
    ): LocationMergeResult
}

File: Sources/RTLSCore/SyncMergeStrategy.swift
import Foundation

/// Context passed to merge strategy when resolving a conflict.
public struct MergeContext: Sendable {
    public var lastSyncAt: Date?
    public var serverTime: Date?

    public init(lastSyncAt: Date? = nil, serverTime: Date? = nil) {
        self.lastSyncAt = lastSyncAt
        self.serverTime = serverTime
    }
}

/// Result of resolving a conflict: keep local, keep server, or use a merged item.
public enum LocationMergeResult: Sendable {
    case keepLocal
    case keepServer
    case use(LocationPoint)
}

/// Optional. If not provided, default is server-wins for pull (overwrite local with server).
public protocol LocationMergeStrategy: Sendable {
    /// Called when local and server both have an item with the same id.
    /// Return the resolved result to persist.
    func resolve(
        local: LocationPoint?,
        server: LocationPoint,
        context: MergeContext
    ) -> LocationMergeResult
}
